### PR TITLE
Reduce duplicate specs

### PR DESCRIPTION
**Pull request by mbj**
* `current_order` cannot return a completed order, by definition

Extraction from #398 to make sure changes to specs needed by the shipment removal do not cause redundant work.

### CODE REVIEW

**Comment by mbj** [comment]
@dkubb This is ready for review.

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -54,6 +54,20 @@ steps:
...
 
+  - name: ':skull: Spree::CheckoutController (parts)'
+    command: |
+      docker-wrapper --path frontend --database -- bash -c '
+        ../scripts/setup-ci.sh frontend
+        bundle exec mutant --use rspec --require ./spec/dummy/config/environment --since HEAD~1 -- \
+         Spree::CheckoutController#update                                                          \
+         Spree::CheckoutController#post_update_redirect_path                                       \
+         Spree::CheckoutController#render_unprocessable_entity                                     \
+         Spree::CheckoutController#rescue_from_spree_gateway_error
```
Comment (dkubb) [inline .buildkite/pipeline.yml]: What about using `Spree::CheckoutController*` and then using `--ignore-subject` to exclude those parts which don't yet pass with mutant?

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 2]
```diff
@@ -32,26 +31,59 @@ def edit
...
     # Updates the order and advances to the next state (when possible.)
     def update
-      if @order.update_from_params(params, permitted_checkout_attributes, request.headers.env)
-        @order.temporary_address = !params[:save_user_address]
-        unless @order.next
-          render :edit, status: :unprocessable_entity and return
-        end
+      unless @order.update_from_params(params, permitted_checkout_attributes, env)
+        render_unprocessable_entity
+        return
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: What if we did `return render_unprocessable_entity` by convention?

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 3]
```diff
@@ -32,26 +31,59 @@ def edit
...
-        else
-          redirect_to checkout_state_path(@order.state)
-        end
-      else
-        render :edit, status: :unprocessable_entity
+      @order.temporary_address = params[:save_user_address].eql?('1')
+
+      unless @order.next
+        render_unprocessable_entity
+        return
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/471/files#r74180074

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (mbj) [inline .buildkite/pipeline.yml]: Its for me right now easier to enumerate (and keep adding to it) what I explicitly touched, especially on a high entropy low coverage and legacy thing the checkout controller is.

I think inversing the strategy is best in the next iteration if not that iteration reaches 100% coverage.

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 4]
```diff
@@ -1,4 +1,144 @@
-describe Spree::CheckoutController, type: :controller, db: :transaction do
+describe Spree::CheckoutController, type: :controller do
+  let(:order) { mock_model(Spree::Order, state: 'delivery', to_param: 'R1234567890') }
+
+  describe '#update' do
+    let(:post_params)    { { state: 'address' } }
+    let(:update_success) { true                 }
+
+    def do_request
+      spree_post :update, post_params.dup
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: What is the reason for the `#dup` here?

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 5]
```diff
@@ -1,4 +1,144 @@
...
+    end
+
+    before do
+      expect(controller).to receive_messages(
+        current_order:              order,
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: This value is not aligned with the other values in the Hash.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 2 above)
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: Its common in rails actions to "bundle" method calls with control flow. But in all non rails code we write (and that is the majority) we'd not do this.

File: frontend/spec/controllers/spree/checkout_controller_spec.rb (hunk 4 above)
Comment (mbj) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: `spree_post` mutates the params in-place. Making them un-reusable in most of the cases for other specs.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 2 above)
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: I think in general we usually follow the convention of returning the result of the `render(...)` call in most rails actions. We don't tend to short-circuit an action like this though, which is why it looks a bit weird to write (effectively( `return render(...)`.

Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: Its not a render call. Its a wrapper around a render call. Is there the equivalent of `halt` we can call from `render_unprocessable_entity` ?

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (dkubb) [inline .buildkite/pipeline.yml]: Can the subjects be sorted alphabetically so they are added in a deterministic order?

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 2 above)
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: > Its not a render call. Its a wrapper around a render call.

I said it's _effectively_ `return render(...)`, which is what we do in almost all controller actions we write.

> Is there the equivalent of halt we can call from render_unprocessable_entity ?

I don't think there is.

Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: I can agree with using it for now.

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 6]
```diff
@@ -1,4 +1,144 @@
-describe Spree::CheckoutController, type: :controller, db: :transaction do
+describe Spree::CheckoutController, type: :controller do
+  let(:order) { mock_model(Spree::Order, state: 'delivery', to_param: 'R1234567890') }
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: If you replace `mock_model` with `instance_double` does it still work? I think if we can we should try to do:

```
lightweight value object > verified double > mock model > real model
```

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 7]
```diff
@@ -1,4 +1,144 @@
...
+        ensure_checkout_allowed:       nil,
+        ensure_sufficient_stock_lines: nil,
+        set_state_if_present:          nil
+      )
+
+      allow(order).to receive(:update_from_params) do |controller_params, permitted_attributes, rack_env|
+        expect(controller_params).to include(post_params)
+        expect(permitted_attributes).to eql(controller.permitted_checkout_attributes)
+        expect(rack_env).to be(controller.env)
+      end.and_return(update_success)
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: Is the `and_return` needed here? Can't the block just return `update_success` ?

File: frontend/spec/controllers/spree/checkout_controller_spec.rb (hunk 6 above)
Comment (mbj) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: I like that one.

**Comment by mbj** [comment]
@dkubb This is ready for the next pass!

File: frontend/spec/controllers/spree/checkout_controller_spec.rb (hunk 6 above)
Comment (mbj) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: In this case `mock_model` is required as model specific API is used (`#errors`).

**Comment by dkubb** [comment]
@mbj I had one comment in https://github.com/MountainRoseHerbs/spree/pull/471#discussion_r74180693 that was pending.

**Comment by mbj** [comment]
> @mbj I had one comment in #471 (comment) that was pending.

After pinging you for next review I discovered some new alive mutations and pushed an experiment to determine if the move of the `update_success` location makes a difference. You observed that state without my intention at the time you created this comment.

Long story short: Its fixed by now.

**Pull request merged**